pandas/core/algorithms.py

```diff
@@ -1678,9 +1678,23 @@ def map_array(
     if not len(arr):
         return arr.copy()
 
-    # we must convert to python types
-    values = arr.astype(object, copy=False)
+    na_value = None
+    mask = isna(arr)
+    if isinstance(arr.dtype, BaseMaskedDtype) and na_action is None:
+        arr = cast("BaseMaskedArray", arr)
+        values = arr._data
+        if arr._hasna:
+            na_value = arr.dtype.na_value
```

Comment:
It seems to me we want to be doing this for all EAs, not just masked ones.

Reply:
@rhshadrach: Hello Richard, thanks for your feedback. I'm agree it is a partial solution to the issue (a solution limited to `BaseMaskedArray`).

One solution could be to replace the following code in the function `map_array`:
```
    if isinstance(arr.dtype, BaseMaskedDtype) and na_action is None:
        arr = cast("BaseMaskedArray", arr)
        values = arr._data
        if arr._hasna:
            na_value = arr.dtype.na_value
    else:
        # we must convert to python types
        values = arr.astype(object, copy=False)

```

by this one (taking into account `ExtensionArray`):

```
    if isinstance(arr.dtype, BaseMaskedDtype) and na_action is None:
        arr = cast("BaseMaskedArray", arr)
        values = arr._data
        if arr._hasna:
            na_value = arr.dtype.na_value
    elif isinstance(arr.dtype, ExtensionDtype) and na_action is None:
        arr = cast("ExtensionArray", arr)
        values = np.asarray(arr)
        if arr._hasna:
            na_value = arr.dtype.na_value
    else:
        # we must convert to python types
        values = arr.astype(object, copy=False)
```

Nevertheless, this code just extracts from an array:
- its data as a ndarray (I mean data different from na values)
- its mask as a ndarray
- its na value (`pd.NA`, `np.nan`, ...)

So, maybe this pattern (if I understood correctly what Will was saying :-)) is already managed somewhere else in the code (or I could adapt it) ?

Reply:
> So, maybe this pattern (if I understood correctly what Will was saying :-)) is already managed somewhere else in the code (or I could adapt it) ?

Not that I am aware of, but I might just not be aware.